### ISSUE TITLE
Jodit is not a constructor

### ISSUE THREAD

**Issue by Failender** [jodit/jodit-angular#25]
I followed the steps to integrate jodit for an angular application (using angular7).
Sadly it is not working and exists with the exception "Jodit is not a constructor".
I created a easy repro stackblitz here:
https://stackblitz.com/edit/angular-swuqxr?file=src%2Fapp%2Fapp.component.html

**Comment by hachadwick** [comment]
I am experiencing the same exact issue.

**Comment by hachadwick** [comment]
I rolled back to the jodit-angular 1.0.79 and I don't have the same issue

**Comment by hachadwick** [comment]
just for testing purposes, I rolled back to 1.0.77 and it has the same issue with the constructor. 1.0.78 does not

**Comment by hachadwick** [comment]
Mine fails when trying to run this line of code in the jodit-angular/esm5/jodit-angular.js file: `        
  this.editor = new Jodit(this.element, this.config);`

**Comment by hachadwick** [comment]
So, I have nailed this down to this. If I use a older version of Jodit with jodit-angular, this worked fine. This is my package.json: `    "jodit": "^3.2.27",
    "jodit-angular": "^1.0.78",`

I rolled back the jodit version and it fixed the problem.

**Comment by Failender** [comment]
I can confirm that "fix"

**Comment by hachadwick** [comment]
I have some further details that may help some with another little issue related to this one:
When I installed the altest updates the buttons in this section of the editor stopped working correctly: 
![image](https://user-images.githubusercontent.com/32000375/51202256-f0412000-18c3-11e9-9bf3-09bb467a69d0.png).
After some digging, I noticed that it was placing the HTML tags in the wrong place in the code behind the editor. For example, if I click bold, it would place the strong in front of my text, instead of around it. 

Interestingly, the issue with he constructor is also corrected with the following fix and allow for usage of the latest jodit-angular package (1.0.81).

To fix this issue, I did the following:

1. npm install jodit-angular@1.0.81
2. npm **uninstall** jodit
3. npm install jodit@3.2.27

with this configuration, the latest jodit-angular works with that specific version of Jodit.  I hope this is clear and I hope to see a fix int he future. I think the real issue with both problems lies in the Jodit TS application.

**Comment by 6matko** [comment]
Your solution worked for me, thanks.

**Comment by rafacandev** [comment]
I also confirm that @hachadwick solution also worked for me. Here is my setup:
```
Angular CLI: 7.2.2
Node: 10.14.1
OS: linux x64
Angular: 7.2.1

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.12.2 (plus  other angular 0.12.2 dependencies)
@angular-devkit/core              7.2.2    (plus other angular 7.2.2 dependencies)
@schematics/update                0.12.2
typescript                        3.2.4
webpack                           4.28.4
```

**Comment by Scobee** [comment]
@hachadwick @rafasantos @6matko  - I am having some issues with jodit and maybe you can provide a bit of help
For some reason, in the `jodit_angular` folder, there is a `node_modules` folder .. which crashes everything.
If I remove `node_modules` everything works, but if not .. the app crashes.

And the second thing .. did you guys managed to make brush,image or link icons work on mobile ? I have an open question since Oct for this .. and nobody answered: https://github.com/xdan/jodit/issues/96

**Comment by rafacandev** [comment]
Hi @Scobee 
I did not experienced the issues you described. Neither the `node_modules` nor the icon problems (but I only tested on Chrome simulating a mobile device).

I find that steps to reproduce are an excellent way to narrow down issues specially when dealing with problems that might be related to version/dependencies. Here are the steps that worked for me; hopefully that helps you. FYI: I used a clean virtual machine instance to avoid external factors:

```
# Install node v10.0.0
curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.34.0/install.sh | bash
nvm install 10.0.0
# Install angular cli globally
npm install -g @angular/cli
ng --version 
#>  Angular CLI: 7.2.2
# Create a new angular project named 'jodit-angular-sample'
ng new jodit-angular-sample --routing false --style scss --skip-install true
cd jodit-angular-sample
npm install
# Install jodit-angular dependency
npm install jodit-angular@1.0.81

# Add jodit-editor to app.module and the welcome page
sed -i "3iimport { JoditAngularModule } from 'jodit-angular';" src/app/app.module.ts 
sed -i "13i,JoditAngularModule" src/app/app.module.ts
sed -i "1i<jodit-editor></jodit-editor>" src/app/app.component.html
npm start
# At this point, go to http://localhost:4200 it has a console error:
# TypeError: Jodit is not a constructor...

# Hit ctrl+c to stop npm 
npm uninstall jodit
npm install jodit@3.2.27
npm start

# At this point, go to http://localhost:4200 it works fine
```
Here is how it looks:
![jodit-angular-issue-25](https://user-images.githubusercontent.com/5315487/51442290-79cc6580-1ca9-11e9-8981-2cfa1cd0fad8.png)

**Comment by 6matko** [comment]
@Scobee Hello. I didn't experience those issues either. I just tested jodit link via my phone and everything was fine. Also, there really is node_modules folder and I didn't remove it, so probably problem is somewhere else. In your other post you mentioned that you are using Ionic. Have you tried Jodit on empty project, maybe it is somehow related with Ionic ?

**Comment by Scobee** [comment]
Thank you @rafasantos  and @6matko  - I'll try on a clean new project with the same specs as the one I'm using right now, and I'll come back with an answer. 
Currently I am using Angular 5 and Ionic 3 (I saw that you guys are on Angular 7 .. so that might be one of the issues), but I'll check on the empty project.

**Comment by Shrewd** [comment]
I had the "jodit is not a constructor" issue, with @hachadwick fix, it works.

Working versions: 

jodit": "^3.2.27"
jodit-angular": "^1.0.81"

Not working:
jodit": "^3.2.36"
jodit-angular": "^1.0.81"

**Comment by ronfogel** [comment]
If you want to get it working with the latest version follow these steps:

- open `angular.json` file and insert a new entry into the `styles` array

```
"styles": [
  "styles.css",
  "node_modules/jodit/build/jodit.min.css",
]
```

- in `angular.json` file insert a new entry into the `scripts` array

```
"scripts": [
  "node_modules/jodit/build/jodit.min.js",
]
```

**Comment by hachadwick** [comment]
> If you want to get it working with the latest version follow these steps:
> 
> * open `angular.json` file and insert a new entry into the `styles` array
> 
> ```
> "styles": [
>   "styles.css",
>   "node_modules/jodit/build/jodit.min.css",
> ]
> ```
> * in `angular.json` file insert a new entry into the `scripts` array
> 
> ```
> "scripts": [
>   "node_modules/jodit/build/jodit.min.js",
> ]
> ```

Thank you for the post! I agree that this is a much better solution to the problem. Maybe the documentation for the NPM package needs this added to it.

**Comment by hachadwick** [comment]
> If you want to get it working with the latest version follow these steps:
> 
> * open `angular.json` file and insert a new entry into the `styles` array
> 
> ```
> "styles": [
>   "styles.css",
>   "node_modules/jodit/build/jodit.min.css",
> ]
> ```
> * in `angular.json` file insert a new entry into the `scripts` array
> 
> ```
> "scripts": [
>   "node_modules/jodit/build/jodit.min.js",
> ]
> ```

At first I tried this and it worked, but then I realized I did not have the older versions completely removed. I then purged the entire solution of node packages and started from scratch. I then applied this fix and I am still getting a constructor error. Using the older versions of the package works fine. I will continue to test this fix and see if I can get it working.

**Comment by hachadwick** [comment]
I have successfully been able to get the solution @ronfogel proposed to work! It works nicely. Thank you for the post. I believe this entire issue can be closed and marked as closed.

**Comment by dexster** [comment]
Were the packages rebuilt before publishing to npm? When I look at the UMD or ESM dist bundles I don't see the "runOutsideAngular" and I still get that constructor error
```
JoditAngularComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        if (!this.element) {
            this.createElement();
        }
        this.editor = new Jodit(this.element, this.config);
        if (this.defaultValue) {
            this.editor.value = this.defaultValue;
        }
```

When I package the code locally I see this and my project runs fine:

```
JoditAngularComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        if (!this.element) {
            this.createElement();
            this.ngZone.runOutsideAngular(function () {
                _this.editor = new EditorModule.Jodit(_this.element, _this.config);
            });
        }
```

**Comment by yuriykuzin** [comment]
> I have successfully been able to get the solution @ronfogel proposed to work! It works nicely. Thank you for the post. I believe this entire issue can be closed and marked as closed.

@hachadwick How did you manage it? On last versions I still see the same issue ('not a constructor').

**Comment by Raiper34** [comment]
I have same issue and no solutions mentioned here did not worked for me...

**Comment by AAniculaesei-modus** [comment]
@Raiper34 - what Angular version are you using ? I am getting this on Angular 5

**Comment by Ross-Blakeney** [comment]
> @Raiper34 - what Angular version are you using ? I am getting this on Angular 5

Yeah. I'm stuck on angular 5 as well. According to this issue (https://github.com/jodit/jodit-angular/issues/24) it seems like there are other issues around angular 5 but I haven't had any success in figuring it out yet.

**Comment by AAniculaesei-modus** [comment]
@Raiper34 - I am using "jodit": "^3.2.27" and "jodit-angular": "^1.0.81" in my app, with Angular 5 .. the only issue is that I have to delete node_modules from the jodit director after every single npm install
Another issue I saw is that the links / images buttons don't work ..

**Comment by Ross-Blakeney** [comment]
I can't confirm if @ronfogel's solution works since my app uses webpack and I couldn't get that working similarly.

My "solution" was to use these explicit versions:
"jodit": "3.2.10",
"jodit-angular": "1.0.59"

I tried for hours to get later versions of jodit-angular working in my angular 5 app and this is the only set of versions I can get working. Note: I had to specify the jodit version because the nested dependency in jodit-angular installs a version of jodit that is _seemingly_ not compatible. Any version of jodit higher than this gives the "jodit not a constructor" error.

**Comment by entith** [comment]
@xdan Would you be able to publish a new release of this package that includes the fix for this from #26?

For everyone else, I was able to get this working by using the current latest version of `jodit-angular` (^1.0.81) and rolling the `jodit` package back to ^3.2.31, as that seems to be the last version before the exports were changed.

**Comment by giancafigueroa** [comment]
#34 I had this error but I researched and fix like #34

**Comment by aniketrk** [comment]
Guys I have one issue I am using this editor in Angular 8 version. I am trying to pass html snippet through editor. But some how editor converts the element attributes to lower case for Example
`<div [ngSwitch]="x"></div>` is transformed to `<div [ngswitch]="x"></div>`
`<p [routerLink]="['/error']"></p>` is transformed to `<p [routerlink]="['/error']"></p>`

I even tried with **cleanhtml** but even that didn't help.

Have anyone faced this issue? 
Do we have any config option to avoid this?

Thanks In Advance.
